docs(recipes): Pagination (#3284)
cube-js/cube.js: docs/content/Examples-Tutorials-Recipes/Pagination.md

---
title: Pagination
permalink: /recipes/pagination
category: Examples & Tutorials
subCategory: Queries
menuOrder: 2
---

> ## Use case

Reviewer: Maybe having an example of the data schema would be helpful?

Author: I have added the data schema to the `Data schema` section

> We want to display a table of data with hundreds of rows. For the convenience of

Reviewer 2: ```suggestion
We want to display a table of data with hundreds of rows.
```

> displaying, we will use a pagination. At the recipe below, we will get the

Reviewer 2: ```suggestion
To make the table easier to digest and to improve the performance of the query, we'll use pagination. With the recipe below, we'll get the
```

orders list sorted by the order number. Every page will have 5 orders.

## Query

To creating the pagination we can use the `limit` and `offset` query properties.
First, we get the number of all orders that we have. Then we should set the
`limit` and `offset` properties for the queries that will get the orders from

> Cube.js API.

Reviewer 2: ```suggestion
the Cube.js API.
```

```bash
// Get count of the orders
curl cube:4000/cubejs-api/v1/load \

> -G -s --data-urlencode "query={"measures": ["Orders.count"], "timeDimensions": [{"dimension": "Orders.createdAt"}]}"

Reviewer: Do we need timeDimension here?

Author: Nice catch! We don't need timeDimension here. I deleted it.

```

```bash
// Get first five orders
curl cube:4000/cubejs-api/v1/load \
-G -s --data-urlencode "query={"order": [["Orders.number", "asc"]], "dimensions": ["Orders.number"], "limit": 5}"
```

```bash
// Get next five orders
curl cube:4000/cubejs-api/v1/load \
-G -s --data-urlencode "query={"order": [["Orders.number", "asc"]], "dimensions": ["Orders.number"], "limit": 5, "offset": 5}"
```

## Result

> We have received five orders per query and can use it as we want.

Reviewer 2: ```suggestion
We have received five orders per query and can use them as we want.
```

> ```javascript

Reviewer 3: ```suggestion
```json
```

Reviewer 4: @hassankhan Do you like how JSON is highlighted better than JavaScript? 😄

// Orders count:

[
  {
    'Orders.count': '10000',
  },
];
```

> ```javascript

Reviewer 3: ```suggestion
```json
```

Author: Yeah, I wanted to use `json`, but I thought it looked nicer with `javascript`.

`javascript`
<img width="1552" alt="Screenshot 2021-08-20 at 10 15 23" src="https://user-images.githubusercontent.com/18264667/130183209-e5905d40-cbe7-4aac-9111-c6c7321d5388.png">

`json`
<img width="1552" alt="Screenshot 2021-08-20 at 10 15 47" src="https://user-images.githubusercontent.com/18264667/130183243-40fd537f-9c80-47b4-ab41-7289c2d7ede9.png">

What do you think?

Reviewer 4: I'd stick to JavaScript 😄

// The first five orders:

[
  {
    'Orders.number': 1,
  },
  {
    'Orders.number': 2,
  },
  {
    'Orders.number': 3,
  },
  {
    'Orders.number': 4,
  },
  {
    'Orders.number': 5,
  },
];
```

> ```javascript

Reviewer 3: ```suggestion
```json
```

---
title: Pagination
permalink: /recipes/pagination
category: Examples & Tutorials
subCategory: Queries
menuOrder: 2
---

## Use case

We want to display a table of data with hundreds of rows. To make the table
easier to digest and to improve the performance of the query, we'll use
pagination. With the recipe below, we'll get the orders list sorted by the order
number. Every page will have 5 orders.

## Data schema

We have the following data schema.

```javascript
cube(`Orders`, {
  sql: `SELECT * FROM public.orders`,

  measures: {
    count: {
      type: `count`,
    },
  },

  dimensions: {
    number: {
      sql: `number`,
      type: `number`,
    },

    createdAt: {
      sql: `created_at`,
      type: `time`,
    },
  },
});
```

## Query

> To creating the pagination we can use the `limit` and `offset` query properties.

Reviewer 4: ```suggestion
To select rows that belong to a particular page, we can use the `limit` and `offset` query properties.
```

First, we get the number of all orders that we have. Then we should set the
`limit` and `offset` properties for the queries that will get the orders from

> the Cube.js API.

Reviewer 4: ```suggestion
the Cube API.
```

Let's say Cube instead of Cube.js, when possible (when there's no confusion).

```bash
// Get count of the orders
curl cube:4000/cubejs-api/v1/load \

> -G -s --data-urlencode "query={"measures": ["Orders.count"]}"

Reviewer 4: I think you can just show the queries, all this `curl...` stuff just makes it harder to understand what's going on.
